packages/react-router/modules/resolveLocation.js

```diff
@@ -0,0 +1,38 @@
+import resolvePathname from 'resolve-pathname'
+
+const resolveLocation = (location, base) => {
+  base = !base ? '/' : base
+  if (location === undefined) {
+    return base
+  } else if (typeof location === 'string') {
+    return join(location, base)
+  } else {
+    location.pathname = join(location.pathname, base)
+    return location
+  }
+}
+
+const join = (to, from) => {
+  if (isAbsolute(to)) {
+    return to
+  }
+  const toSlash = hasTrailingSlash(to)
+  const fromSlash = hasTrailingSlash(from)
+  const pathname = resolvePathname(to, withTrailingSlash(from))
+  return sameSlash(pathname, !toSlash && !fromSlash)
+}
+
+const isAbsolute = pathname => !!(pathname && pathname.charAt(0) === '/')
+
+const withTrailingSlash = pathname =>
+  hasTrailingSlash(pathname) ? pathname : pathname + '/'
+
+const hasTrailingSlash = pathname => 
+  !!pathname && pathname.charAt(pathname.length-1) === '/'
+
+const sameSlash = (pathname, noSlash) => noSlash ? stripTrailingSlash(pathname) : pathname
```

Comment:
Yes, the name is pretty poor. It is something I should probably just add some comments on to clarify. The situation that this exists for is when the base does not have a trailing slash and the pathname is an empty string or uses dot notation.
```js
resolvePathname('', '/base')
```
That creates the URL segments `['', 'base', '']` which join to create `/base/`, so I look at the trailing slashes on the original paths to determine whether the returned pathname has one.